_posts/2019-02-11-three-useful-data-migrations-patterns-in-rails.markdown

```diff
@@ -0,0 +1,249 @@
+---
+layout: post
+title: "Three Useful Data Migration Patterns for Rails"
+date: 2019-02-05 10:00:00
+categories: ["rails", "data-migrations"]
+author: "etagwerker"
+---
+
+At [Ombu Labs](https://www.ombulabs.com), we are big fans of [Ruby on Rails](https://rubyonrails.org) and [design patterns](https://www.ombulabs.com/blog/tags/design-patterns), especially [convention over configuration](https://rubyonrails.org/doctrine/#convention-over-configuration)! The beauty of Rails is that
+you can inherit a legacy project and easily find the different layers of code
+in different directories.
+
+When it comes to database migrations the policy of Rails is very clear. It's all
+about altering the database structure with gradual migration files:
+*"Migrations are a convenient way to alter your database schema over time in a
+consistent and easy way."* ([source](https://edgeguides.rubyonrails.org/active_record_migrations.html#migration-overview))
+
+But, what about data migrations? What's the best way to write, maintain, and run
+migrations that alter the data in your production database?
+
+In this article I will talk about three different patterns for writing and
+maintaining your data migrations:
+
+1. Data migrations in `db/migrate`
+2. Data migrations with a set of Rake tasks
+3. Data migrations with `data_migrate`
+
+<!--more-->
+
+## Data migrations in `db/migrate`
+
+When you are starting with Ruby on Rails, it seems like a good idea to add data
+migrations to your database migrations. You can quickly take advantage of
+core Rails functionality and start modifying records across environments.
+
+Let's say that you need to change the default state of a blog post, you can
+create a migration file like this:
+
+```
+rails generate migration ChangeDefaultState
+  invoke  active_record
+  create  db/migrate/20190211222539_change_default_state.rb
+```
+
+In that file, you can write a statement like this one:
+
+```
+class ChangeDefaultState < ActiveRecord::Migration[5.1]
+  def up
```

Comment:
@etagwerker I think you're missing the line to set the new default value: `change_column_default(:posts, :state, "TODO")`

Reply:
@lubc I decided to keep it simple. I could do that, but it's not the point of this article.